### ISSUE TITLE
Keys (Tab, arrows) should not be handled by RenderEditable when also processed as an action.

### ISSUE THREAD

**Issue by stuz5000** [flutter/flutter#47156]
In most UIs, TAB can be input to switch between input fields. This is necessary to enable accessible applications and also application that behave correctly with an external keyboard. This behavior doesn't typically ALSO input tab into the input field. In flutter tab performs both:

 * switch focus
 * input whitespace text into text fields.

If the first is done, the second should not.

```
import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(
      title: 'My app', // used by the OS task switcher
      home: Material(child: content())));
}

Widget content() {
  var _focus1 = FocusNode();
  var _focus2 = FocusNode();

  return Column(
    children: [
      TextField(
        focusNode: _focus1,
        textInputAction: TextInputAction.next,
      ),
      TextField(
        focusNode: _focus2,
        textInputAction: TextInputAction.next,
      ),
    ],
  );
}
```

Is with the latest public flutter, Android emulator and Nexus 7 tablet VM.

<details>
<summary> flutter doctor -v</summary>

```
[✓] Flutter (Channel master, v1.13.3-pre.23, on Linux, locale en_US.UTF-8)
    • Flutter version 1.13.3-pre.23 at /home/sir/development/flutter
    • Framework revision c06bf6503a (3 days ago), 2019-12-13 17:42:35 -0500
    • Engine revision e0e0ac0a68
    • Dart version 2.8.0 (build 2.8.0-dev.0.0 45db297095)

 
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/sir/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /home/sir/Android/Sdk
    • Java binary at: /home/sir/soft/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /home/sir/soft/android-studio
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.3)
    • IntelliJ at /snap/intellij-idea-ultimate/188
    • Flutter plugin version 42.1.4
    • Dart plugin version 193.5731

[✓] IntelliJ IDEA Ultimate Edition (version 2019.2)
    • IntelliJ at /snap/intellij-idea-ultimate/181
    • Flutter plugin version 41.1.4
    • Dart plugin version 192.7761

[✓] Connected device (3 available)
    • Android SDK built for x86 • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Chrome                    • chrome        • web-javascript • Google Chrome 76.0.3809.132
    • Web Server                • web-server    • web-javascript • Flutter Tools

! Doctor found issues in 1 category.
```
</details>

**Comment by HansMuller** [comment]
CC @gspencergoog

**Comment by gspencergoog** [comment]
In order to do this properly, we probably need to have https://github.com/flutter/flutter/issues/33521 implemented, since otherwise we can't have the Framework block insertion of a key because it is handling it as a focus signal (or any other shortcut).

**Comment by gspencergoog** [comment]
This doesn't just affect tab keys: it affects arrow keys too if you tell the DirectionalFocusIntent to not ignore text fields.

Really, it affects any key bound to another action.  The action should absorb the key, and it should never be delivered to the text field.

**Comment by gspencergoog** [comment]
In order to solve this, a rather large structural change is needed in the way Flutter processes key events.  

Here's the design doc for moving the key infrastructure to the platform code: https://flutter.dev/go/platform-based-key-events

The first part of this is underway already. https://github.com/flutter/flutter/pull/50998 and https://github.com/gspencergoog/platform_keyboard are work towards the goal of being able to synchronously handle key events and cancel propagation of the event when it has been handled. The first is a PR that will allow us to generate key codes in C++ for the various platforms, and the second will become the framework-side changes for the new synchronous keyboard API.

That will fix this issue by allowing key events to have their propagation stopped when an action handles them.

**Comment by gspencergoog** [comment]
I'm currently working through adding synchronous key event support to Android as a proof of the design, and will move on to other platforms once that lands.

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.21.0-2.0.pre.66` on a Pixel 3 emulator using Android API 30.

<details>
 <summary>flutter doctor -v</summary>

 ```
[√] Flutter (Channel master, 1.21.0-2.0.pre.66, on Microsoft Windows [Version 10.0.18362.900], locale et-EE)
    • Flutter version 1.21.0-2.0.pre.66 at C:\Development\flutter_master
    • Framework revision 54c9441723 (3 days ago), 2020-07-17 18:56:12 -0400
    • Engine revision d327828632
    • Dart version 2.9.0 (build 2.9.0-21.0.dev 9dca49e71e)


[√] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at C:\Users\Isa\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30309.148
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.0)
    • Android Studio at C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.47.2)
    • VS Code at C:\Users\Isa\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.12.2

[√] Connected device (6 available)
    • SM G950F (mobile)           • ce12171c51cc001c03 • android-arm64  • Android 9 (API 28)
    • sdk gphone x86 (mobile)     • emulator-5554      • android-x86    • Android 11 (API 30) (emulator)
    • sdk gphone x86 arm (mobile) • emulator-5556      • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)           • windows            • windows-x64    • Microsoft Windows [Version 10.0.18362.900]
    • Web Server (web)            • web-server         • web-javascript • Flutter Tools
    • Chrome (web)                • chrome             • web-javascript • Google Chrome 84.0.4147.89

• No issues found!
```
</details>

**Comment by gspencergoog** [comment]
I landed synchronous key event support for Android, but then had to revert it because it broke some benchmarks that were using the back button.  I'm currently diagnosing that so I can re-land it.

We're also working on a design for editing actions to be implemented as Flutter Actions so that the shortcuts framework can trigger them, which will solve some of the precedence issues.

**Comment by znjameswu** [comment]
> I landed synchronous key event support for Android, but then had to revert it because it broke some benchmarks that were using the back button. I'm currently diagnosing that so I can re-land it.
> 
> We're also working on a design for editing actions to be implemented as Flutter Actions so that the shortcuts framework can trigger them, which will solve some of the precedence issues.

I saw in that PR it is stated that typically only hardware keyboard event is captured. Is it possible to extend that to soft keyboard events as well? It could be done by overrideing `onKeyPreIme` method on Android, but I don't know for other platforms. I'm looking forward to that feature.

**Comment by gspencergoog** [comment]
I was able to re-land the event synthesis for Android.

@znjameswu I think the `onKeyPreIme` method still only sends hardware key events, it just sends them before the IME gets them. I haven't found a way to capture soft key events before the IME gets them on Android, only hardware key events.

I suspect that there isn't any such API, because of this documentation from the [KeyEvent](https://developer.android.com/reference/android/view/KeyEvent) Android docs:
> As soft input methods can use multiple and inventive ways of inputting text, there is no guarantee that any key press on a soft keyboard will generate a key event: this is left to the IME's discretion, and in fact sending such events is discouraged. You should never rely on receiving KeyEvents for any key on a soft input method. In particular, the default software keyboard will never send any key event to any application targetting Jelly Bean or later, and will only send events for some presses of the delete and return keys to applications targetting Ice Cream Sandwich or earlier. Be aware that other software input methods may never send key events regardless of the version.

**Comment by gspencergoog** [comment]
So, just to update: I have prototyped a way to do this on Android, and I think I can do it on macOS, Windows and Linux, although I haven't yet prototyped those.

Unfortunately, on Android, it only works for hardware keyboards (as alluded to above), so pressing a key that is handled by a shortcut on the hardware keyboard can be intercepted and handled without passing it on to the text field, but if the same key is pressed on a software keyboard, the key is not handled (and not even sent to the `RawKeyboard` handler, or the `Shortcuts` widget).

The only (super hacky) way I can see to make it work for soft keyboards too is to watch the edit stream and simulate key events based on the delta changes to the text. Maybe it would "just work" most of the time, but I suspect that there will be many cases where it doesn't (e.g. what key event do I send when someone inserts an emoji?). And some editing operations will result in weird things: If the cursor was between the first and second character and I hit "Home", in the edit stream it would be indistinguishable from hitting "Left Arrow", so if the shortcut was for "Home", then hitting "Left Arrow" on the software keyboard might trigger it (but not on the hardware keyboard). And it still wouldn't handle the case of hitting "Backspace" on an empty text field.

Not to mention that it is all very complex code with a lot of assumptions, so I'm not inclined to the hacky solution, but I don't know that we have any other recourse on Android, unless we're fine with shortcuts and keyboard navigation only working for hardware keyboards (which would work for the a11y case that is the original purpose of this issue). That might actually be a reasonable restriction, if a little surprising to developers, given that software keyboards don't typically even contain a "Tab" key or arrow keys to use for focus traversal.

**Comment by znjameswu** [comment]
> I was able to re-land the event synthesis for Android.
> 
> @znjameswu I think the `onKeyPreIme` method still only sends hardware key events, it just sends them before the IME gets them. I haven't found a way to capture soft key events before the IME gets them on Android, only hardware key events.
> 
> I suspect that there isn't any such API, because of this documentation from the [KeyEvent](https://developer.android.com/reference/android/view/KeyEvent) Android docs:
> 
> > As soft input methods can use multiple and inventive ways of inputting text, there is no guarantee that any key press on a soft keyboard will generate a key event: this is left to the IME's discretion, and in fact sending such events is discouraged. You should never rely on receiving KeyEvents for any key on a soft input method. In particular, the default software keyboard will never send any key event to any application targetting Jelly Bean or later, and will only send events for some presses of the delete and return keys to applications targetting Ice Cream Sandwich or earlier. Be aware that other software input methods may never send key events regardless of the version.

Awesome work!

For the Android key event system, I think one discrepency is that native android app can receive a event if a soft keyboard decided to send that event, via `View.addOnKeyListener`. There are some Android IME (such as Hacker's Keyboard) that support extra keys such as 'ctrl', 'alt', etc, and they did pretty well on sending those event. However in my recent testing (1.22.0-10.0.pre.161), `RawKeyListener` or `Focus` seems to be unable to receive those event.

**Comment by gspencergoog** [comment]
@znjameswu I don't see `addOnKeyListener` anywhere in the Android docs. Can you add a link to what you're thinking of? (maybe you mean [`setOnKeyListener`](https://developer.android.com/reference/android/view/View#setOnKeyListener(android.view.View.OnKeyListener))?)

**Comment by znjameswu** [comment]
> @znjameswu I don't see `addOnKeyListener` anywhere in the Android docs. Can you add a link to what you're thinking of? (maybe you mean [`setOnKeyListener`](https://developer.android.com/reference/android/view/View#setOnKeyListener(android.view.View.OnKeyListener))?)

Sorry my typo. It is `setOnKeyListener`. In my testing, it can log various keys from Hacker's Keyboard.

**Comment by gspencergoog** [comment]
Setting the on key listener doesn't appear to get any key events from the Hacker's Keyboard for me. Maybe this has to do with the `FlutterView` being a `ViewGroup`, but I also tried setting it on all children too, and nothing came through.

The Hacker's Keyboard does appear to send some key events through `InputConnection.sendKeyEvent` (but not all, just some specific ones, like TAB and BACKSPACE). It seems to send a LEFT_SHIFT in addition to the TAB when I just press TAB, which is strange.

As I mentioned above, it's up to each IME to decide what key events it will simulate, and I don't think we can count on any particular IME being consistent with any other.

**Comment by znjameswu** [comment]
> Setting the on key listener doesn't appear to get any key events from the Hacker's Keyboard for me. Maybe this has to do with the `FlutterView` being a `ViewGroup`, but I also tried setting it on all children too, and nothing came through.
> 
> The Hacker's Keyboard does appear to send some key events through `InputConnection.sendKeyEvent` (but not all, just some specific ones, like TAB and BACKSPACE). It seems to send a LEFT_SHIFT in addition to the TAB when I just press TAB, which is strange.
> 
> As I mentioned above, it's up to each IME to decide what key events it will simulate, and I don't think we can count on any particular IME being consistent with any other.

Well the Tab behavior is wierd. I'm testing in a native Android app and the tab key seems fine for me. 

If there can't be a universal way to handle IME, then it may be as well good to retain the original intentions of various IME authors (e.g., Hacker's Keyboard does not send events for single ASCII keys, but sends events for combination keys such as ctrl+C), because it is up to them to guarantee the correctness of the end result. (i.e., some keyboards don't send backspace event and modify by `InputConnection`, while Hacker's Keyboard sends backspace events and leaves the modifications to android views)

A hypothetical approach (only my current understanding) would be to perform text-diffing on each platform's `InputConnection` equivalent to infer content insertion/backspace/up/down etc that doesn't get emitted key event, and capture emitted key events from IME as if they come from hardware keyboard. Of course this would be a total change to Flutter's text input codebase, but experiments on small packages can be plausible as soon as IME emitted key event get passed to Flutter.

**Comment by gspencergoog** [comment]
> A hypothetical approach (only my current understanding) would be to perform text-diffing on each platform's InputConnection equivalent to infer content insertion/backspace/up/down etc that doesn't get emitted by key event, and capture emitted key events from IME as if they come from hardware keyboard. Of course this would be a total change to Flutter's text input codebase, but experiments on small packages can be plausible as soon as IME emitted key event get passed to Flutter.

Yes, that's the "hacky" way I mentioned above.  I don't think it's feasible: it's too much of a hack, with too many ambiguous corner cases.

**Comment by Hixie** [comment]
Greg says Android and macOS are done, and he's finishing up Linux and Windows, which is expected to take another few weeks.

I'm moving this to P3 to take it off the weekly-updates triage list, but it remains a high priority.

**Comment by gspencergoog** [comment]
I've hit a snag (and so reverted the Android key synthesis commit), in that now that we're properly processing the events in the shortcuts system, you can no longer type printable characters into a text field if they are also bound to a shortcut, because the shortcut handles them. For instance, if SPACE is bound to activate (as it is on web), then spaces can no longer be typed into a text field. This is obviously not desirable.

I have a solution to this, which is to have a new property on an `Intent` that allows a text field to ask if the intent should be active in a text field, and a new callback registration on `FocusNode` that allows the focus system to ask the node if it should stop key event propagation or not, allowing the text field to stop event propagation, but still notify the engine that the key was unhandled, so that the key can proceed along the text editing/entry path in the engine.

I thought about making it so that text fields just stop propagation for any key event resulting in a printable character, but then the definition of "printable" gets fuzzy: does a newline count as printable? What about a tab? And if a tab is printable, then the shortcuts system can't invoke the action to move to the next field, and we're back to the original problem of getting stuck in text fields when focus traversing via tab.

Having the `Intent` indicate this allows for all possibilities: passing space to text fields, but not tabs, or overriding the default tab binding to allow tabs to be inserted into a text field.

**Comment by gspencergoog** [comment]
I've found a better solution, in #67359, which seems to solve the problem of not being able to insert spaces, etc.

I tried re-landing the pre-IME key event handling, and had to revert it because it disabled the back button in some situations (when the soft keyboard was up).

After discussing it with some Android folks, I have an updated version of that change which will be ready shortly, and handle keys properly, except that it won't be pre-IME, but it doesn't really need to be, it turns out.

**Comment by stuz5000** [comment]
Wouldn't a good solution be to give widgets with a focus node the ability to eat input and prevent it from being passed to parent widgets with focus nodes. (This was the solution used by GTK, way back).
Shortcuts would be handled somewhere closed to the root.

This would solve other issues I have, such as cursor keys used in text areas contained by scrollable areas causing both cursor movement and list scrolling.

**Comment by gspencergoog** [comment]
That is in fact how it works already.

**Comment by stuz5000** [comment]
I don't believe so.
https://api.flutter.dev/flutter/widgets/RawKeyboardListener-class.html
  onKey → ValueChanged<RawKeyEvent>
... results void. Because it returns void, there's no way to indicate that this class consumed this specific event and not provide it to the parent. 

In other UI toolkits
bool onKey(event){
   if (event.ket == SPACE) { ....; return true}   // consumed space
   if (event.ket == ENTER) { ....; return true}   // consumed space
   return false;   // Did not consumed
}

**Comment by gspencergoog** [comment]
No, that API is a low level API for listening to key events, not handling them.

Check out these:
https://api.flutter.dev/flutter/widgets/Focus/onKey.html
https://api.flutter.dev/flutter/widgets/FocusOnKeyCallback.html

You can set a key handler on a Focus widget (or at a lower level, on the FocusNode itself), and returning true will halt key event propagation to other focus nodes, and (once this issue is completely fixed), also halt the propagation to other non-Flutter components in an add-to-app scenario.

**Comment by fanchou** [comment]
Can't use backspace when using TAB on Mac.   #21231

**Comment by fanchou** [comment]
Can't use backspace when using TAB on Mac.  flutter/engine#21231

**Comment by gspencergoog** [comment]
@fanchou Can you give a better description of what you mean so that I can reproduce the problem?

**Comment by fanchou** [comment]
> @fanchou Can you give a better description of what you mean so that I can reproduce the problem?

Sorry, I made a mistake. The correct situation is that I can't use backspace when I turn on the capslock key.

**Comment by venkatd** [comment]
@gspencergoog wanted to confirm if allowing overriding shortcuts for TextFields is on the list for Flutter web is a priority this month month and if it is related to https://github.com/flutter/engine/pull/21231 and https://github.com/flutter/engine/pull/22340

We have a UI like this:
![image](https://user-images.githubusercontent.com/220686/102103377-eb13e500-3dfa-11eb-9257-edb4d476d313.png)
 
The keys we'd like to override are:
- Up/down arrow keys when there are mentionable users
- Tab key or enter key when there is a completion to expand

Right now we have some hacks that save and restore a TextSelection to undo the keyboard movement but we've found there are a lot of edge cases it introduces.

Don't mind running on the latest master for web if necessary as this is one of the few remaining blockers for releasing our app to beta users (planned for mid-January). Thanks!

**Comment by gspencergoog** [comment]
@venkatd This is something we're working on, but the specific case you're talking about will still take some time.

The details are this: While we're no longer handling the key events in the engine, cursor motion and selection are still handled by the text field itself (`EditableText` widget) in Flutter, and because the text field has focus, it receives the events first, and so for the keys it handles, it absorbs the key event before the `Shortcuts` widget (or any other widget) gets a crack at them. It effectively has its own built-in shortcuts that can't be overridden.

We are working on some changes that should turn the editing commands in the text field into actions with key bindings that can be overridden. At that point, your design should be possible. This is a pretty big change with a lot of moving parts, which is why it will take a while (sorry I don't have a specific timeline for you).

/cc @justinmc

**Comment by gspencergoog** [comment]
@fanchou Yes, this is a known problem on macOS (see https://github.com/flutter/flutter/issues/64761)

Fixing this is being worked on as part of implementing https://github.com/flutter/flutter/issues/44918

/cc @dkwingsmt

**Comment by venkatd** [comment]
@gspencergoog no problem, I understand this is not a simple change. 

In the meantime, do you have any recommendations on workarounds/hacks? I see keyboard handling is happening here.
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/rendering/editable.dart#L513-L552 but doesn't look like it's possible to override.

Would it be workable to move the keyboard handling to public methods to allow subclassing TextField/Edtiable? Maybe the ability to pass in `handleKeyEvent` into TextField and returning false in our app code if we want the Editable to ignore the key.

The clean shortcut handling would be lower priority than suppressing certain `RawKeyEvent` messages.

**Comment by gspencergoog** [comment]
@venkatd No, sorry, we probably can't expose those as new public methods: we'd have to support them going forward then, and that's not the API we want to end up with. As I said, this is one of the tasks we're currently working on, so hopefully it won't be too long.

**Comment by venkatd** [comment]
@gspencergoog understood & thanks for the reasoning

**Comment by Hixie** [comment]
Update from @gspencergoog: macOS, Linux, and Android are all done, we're working on the Windows part at the moment.

**Comment by venkatd** [comment]
Great! Was web already done? Or something for the future?

**Comment by justinmc** [comment]
@vankatd This is my [design doc](https://docs.google.com/document/d/1QaVIr1bbOWJGNyyNsB75sXgTDRMuxP268sLGTBNDut0/edit#) for doing things like the arrow keys and tab keys via Shortcuts/Actions/Intents.  I have a POC that has this working on Mac, and all the mapping is overridable, but it will depend on @gspencergoog's work for the full solution.

**Comment by gspencergoog** [comment]
@venkatd @mdebbar I'm looking at web implementation in the next few days. It should be pretty simple to implement, I don't expect any issues.

**Comment by venkatd** [comment]
We've got a flutter web prototype being used internally. It is designed to replace our React web app so we can share more code with our mobile Flutter app.

Our app is pretty text editing heavy. So anything related to text editing--and most importantly shortcuts/intent overrides--would have the highest impact on our customer experience.

Great to hear it's got the teams focus!

**Comment by moradoronline** [comment]
In the web implementation seems to be working. In SImulators on Mac it have this weird behavior.

**Comment by Sunbreak** [comment]
> In the web implementation seems to be working. In SImulators on Mac it have this weird behavior.

Maybe file another issue?
